Nullptr crash in ImageBitmap::createCompletionHandler when GPU process takes too long to initialize

mwyrzykowski opened the pull request:
#### 82d104c1fefcc75f0723565c35978babf4b2af28
<pre>
Nullptr crash in ImageBitmap::createCompletionHandler when GPU process takes too long to initialize
<a href="https://bugs.webkit.org/show_bug.cgi?id=275325">https://bugs.webkit.org/show_bug.cgi?id=275325</a>
&lt;radar://129354725&gt;

Reviewed by Said Abou-Hallawa.

BitmapImage::create returns nullptr when existingImageBitmap-&gt;buffer()-&gt;copyNativeImage()
returns nullptr. This resulted in a nullptr crash.

ImageBuffer::copyNativeImage() will return nullptr if the GPU process takes too long to initialize. This
occurs when the call to waitForAndDispatchImmediately() in RemoteImageBufferProxy::ensureBackend()
takes longer than the GPUP timeout which is 3_s. This reproduces easily in ASAN since process startup
takes longer in an ASAN build.

Fix this by returning a blank image buffer in this case as we do
if bitmapData is nullptr.

* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createCompletionHandler):

* LayoutTests/http/wpt/2dcontext/imagebitmap/nullptrcrash-when-gpu-process-times-out-expected.txt: Added.
* LayoutTests/http/wpt/2dcontext/imagebitmap/nullptrcrash-when-gpu-process-times-out.html: Renamed from LayoutTests/fast/webgpu/nocrash/fuzz-275325.html.
Add regression test.

Canonical link: <a href="https://commits.webkit.org/279924@main">https://commits.webkit.org/279924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e1bdefdb777f1728ed83eecd47d2d2571a833a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58214 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5667 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57236 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41933 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5698 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44486 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3842 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32477 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47588 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25613 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29267 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4930 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3808 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5152 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59804 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30201 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51908 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31332 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51340 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12077 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32347 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31121 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on LayoutTests/fast/webgpu/nocrash/fuzz-275325.html:
```diff
@@ -0,0 +1,45 @@
+<style>
```
@shallawa issue was reduced to this

shallawa replied:
Why is this test put in the webgpu directory? Should not it be in of these directories instead?

1. LayoutTests/http/wpt/2dcontext/imagebitmap/
2. LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/

And why the name is fuzz-275325.html? Can't the name something more descriptive? 

ImageBitmap is not very familiar area to me. So I am learning if there is something specific about its tests.

mwyrzykowski replied:
I placed it in webgpu since it was initially reported against WebGPU, but indeed I don't think it's related. I will move it over to `LayoutTests/http/wpt/2dcontext/imagebitmap/`

mwyrzykowski replied:
will rename from fuzz-275325 (fuzz-[ID of the bugzilla]) to nullptrcrash-when-gpu-process-times-out.html

shallawa commented on LayoutTests/fast/webgpu/nocrash/fuzz-275325.html:
```diff
@@ -0,0 +1,45 @@
+  } else {
+    log('no GC available');
+  }
+}
```
Do we really need this? Does the GC affect the reproducibility of this crash?

mwyrzykowski replied:
Oh no, this should be removed. It was reduced from the fuzzer test, I forgot to remove it

shallawa commented on LayoutTests/fast/webgpu/nocrash/fuzz-275325.html:
```diff
@@ -0,0 +1,45 @@
+  try {
+let video0 = await videoWithData();
+let imageBitmap0 = await createImageBitmap(video0);
+let imageBitmap1 = await createImageBitmap(imageBitmap0);
```
These lines need to be indented.

shallawa commented on LayoutTests/fast/webgpu/nocrash/fuzz-275325.html:
```diff
@@ -0,0 +1,45 @@
+  } catch (e) {
+    log('error');
+    log(e);
+  }
```
I do not think this code throws an exception to be caught. And these log messages are not printed in the expected file. We just care about whether createImageBitmap() crashes or not. So I think the try-catch block and the log messages can be deleted.

shallawa commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -676,6 +676,10 @@ void ImageBitmap::createCompletionHandler(ScriptExecutionContext& scriptExecutio
     }
 
     RefPtr imageForRender = BitmapImage::create(existingImageBitmap->buffer()->copyNativeImage());
```
Why does `ImageBuffer::copyNativeImage()` return nullptr with the above layout test? Please add this info to the commit message.

mwyrzykowski replied:
`RemoteImageBufferProxy::ensureBackend()` times out in ASAN initially, I expanded on the commit message. I confirmed if I increase the timeout with ASAN, the issue does not occur. This timeout could occur for any number of reasons, ASAN just happened to trigger that path

shallawa commented on LayoutTests/fast/webgpu/nocrash/fuzz-275325.html:
```diff
@@ -0,0 +1,45 @@
+<style>
+  :root { background: #102030e0; color: #99ddbbcc; font-size: 15px; }
+</style>
```
I do not think we need this style section.

shallawa commented on LayoutTests/http/wpt/2dcontext/imagebitmap/nullptrcrash-when-gpu-process-times-out.html:
```diff
@@ -0,0 +1,24 @@
+<script>
+globalThis.testRunner?.waitUntilDone();
```
Can't we use `testRunner.dumpAsText();` so the expected file can have only this text

```
Test passes if it does not crash.
```

shallawa commented on LayoutTests/http/wpt/2dcontext/imagebitmap/nullptrcrash-when-gpu-process-times-out.html:
```diff
@@ -0,0 +1,24 @@
+  let imageBitmap1 = await createImageBitmap(imageBitmap0);
+  globalThis.testRunner?.notifyDone();
+};
+</script>
```
If you use `testRunner.dumpAsText();` you may need to add this after the script:

```
<p>Test passes if it does not crash.</p>
```

mwyrzykowski committed:
Nullptr crash in ImageBitmap::createCompletionHandler when GPU process takes too long to initialize
https://bugs.webkit.org/show_bug.cgi?id=275325
<radar://129354725>

Reviewed by Said Abou-Hallawa.

BitmapImage::create returns nullptr when existingImageBitmap->buffer()->copyNativeImage()
returns nullptr. This resulted in a nullptr crash.

ImageBuffer::copyNativeImage() will return nullptr if the GPU process takes too long to initialize. This
occurs when the call to waitForAndDispatchImmediately() in RemoteImageBufferProxy::ensureBackend()
takes longer than the GPUP timeout which is 3_s. This reproduces easily in ASAN since process startup
takes longer in an ASAN build.

Fix this by returning a blank image buffer in this case as we do
if bitmapData is nullptr.

* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createCompletionHandler):

* LayoutTests/http/wpt/2dcontext/imagebitmap/nullptrcrash-when-gpu-process-times-out-expected.txt: Added.
* LayoutTests/http/wpt/2dcontext/imagebitmap/nullptrcrash-when-gpu-process-times-out.html: Renamed from LayoutTests/fast/webgpu/nocrash/fuzz-275325.html.
Add regression test.

Canonical link: https://commits.webkit.org/279924@main
